### PR TITLE
Migrate specs tests to PennyLane #9841 resource API

### PR DESCRIPTION
**Pull request by Qottmann**
## Summary

PennyLane **dev62** is the minimum pin that includes both [PennyLaneAI/pennylane#9846](https://github.com/PennyLaneAI/pennylane/pull/9846) (`magic-T`) and [PennyLaneAI/pennylane#9841](https://github.com/PennyLaneAI/pennylane/pull/9841) (resource API). This PR bumps Catalyst to that pin and fixes the resulting fallout.

**Resource API migration (#9841)**
- Bump PennyLane to `0.46.0.dev62` in `.dep-versions` and `doc/requirements.txt`
- Replace `SymbolicSpecsResources` checks with `SpecsResources.is_symbolic`
- Rename resource fields in tests: `gate_types`/`gate_counts` → `quantum_operations`, `measurements` → `measurement_processes`, `num_gates` → `total_quantum_operations`
- Drop `gate_sizes` assertions (no longer tracked)
- Handle `PBCSpecsResources` in PPR specs tests (depth fields)

**dev62 fallout (Operator2 / magic-T)**
- `decompose.py`: safe `.params` access for Operator2 ops; skip abstract `Adjoint(...)` / `Controlled(...)` decomposition-graph nodes that lack `op_type`
- PLXPR tests: single-qubit ops like `S`/`T` now lower to `qref_operator` instead of `qref_qinst`
- `test_adjoint.py`: replace removed `qp.Rotation` with `qp.Rot`
- `test_quantum_control.py`: compare `ControlledOp2` decompositions with `qp.equal`
- `test_specs.py`: normalize doubly-controlled `C(S)` counts in `test_complex`

This is the base branch for the magic-state allocation stack (`#3029` → `#3027`).

### CODE REVIEW

File: frontend/catalyst/from_plxpr/decompose.py [hunk 1]
```diff
@@ -355,6 +346,26 @@ def _create_decomposition_rule(
...
+            qp.ops.ChangeOpBasis,
+            qp.ops.Prod,
+            qp.TemporaryAND,
+        ),
+    ):
+        return True
+
+    # Operator2 resource reps may omit op_type but still encode symbolic ops in the name.
+    name = op_rep.name
+    return name.startswith(("Adjoint(", "Controlled("))
```
Comment (JerryChen97) [inline frontend/catalyst/from_plxpr/decompose.py]: But I thought ControlledOp2's name has been overriden as `C(...)` ? Would be helpful if @andrijapau can also double check if this will miss things e.g. `C(S)` etc.

File: frontend/test/pytest/test_specs.py
Comment (JerryChen97) [inline frontend/test/pytest/test_specs.py:1]: `check_specs_resources_same()` now accepts either `SpecsResources` or `PBCSpecsResources` on both sides, but only compares the PBC depth fields when both objects are PBC instances.

Because `PBCSpecsResources` subclasses `SpecsResources`, these mismatches pass:

 - actual `SpecsResources`, expected `PBCSpecsResources`;
 - actual `PBCSpecsResources`, expected `SpecsResources`.

The first case is especially problematic: Catalyst could regress to returning ordinary resources and lose `any_commuting_depth` and` qubit_disjoint_depth`, while both new PBC tests still pass. Those fields are the defining additions of the subclass. If this case is truly important we might also need to add exact type equality

**Review by JerryChen97 — CHANGES_REQUESTED** [review]
Also I'm investigating why [previously observed necessary changes](https://github.com/PennyLaneAI/catalyst/pull/3035/changes#diff-9a21c91a4ebeae0a86592c0f5931b932d7882c1a40007c68619a8055030bee4a) are not actually needed for CI...

File: frontend/catalyst/from_plxpr/decompose.py (hunk 1 above)
Comment (Qottmann) [reply frontend/catalyst/from_plxpr/decompose.py]: Yes you are right! Operator2 controlled resource reps use C(S), not Controlled(S). The skip logic now also matches names starting with C(, with a comment explaining why. Verified it skips C(S) but not CNOT/CRX.

Comment (Qottmann) [reply frontend/catalyst/from_plxpr/decompose.py]: ✅

File: frontend/test/pytest/test_specs.py
Comment (Qottmann) [reply frontend/test/pytest/test_specs.py:1]: I turned it to now `assert type(actual_res) is type(expected_res)` so a plain `SpecsResources` actual cant pass against a `PBCSpecsResources`

**Comment by Qottmann** [comment]
> Also I'm investigating why [previously observed necessary changes](https://github.com/PennyLaneAI/catalyst/pull/3035/changes#diff-9a21c91a4ebeae0a86592c0f5931b932d7882c1a40007c68619a8055030bee4a) are not actually needed for CI...

I think those changes are actually necessary! This PR actually mostly avoid (or accept) the qfunc+Operator2 control bug. I pulled in the changes from your PR so the bug gets actually fixed

**Review by JerryChen97 — APPROVED** [review]
Thanks! Let's have a changelog for the source code change (no need for those tests adjustment) and find a Catalyst reviewer

**Comment by Qottmann** [comment]
> Let's have a changelog for the source code change (no need for those tests adjustment)

what would that entail? I dont think there is any user-facing upgrade no?

File: frontend/catalyst/from_plxpr/qfunc_interpreter.py [hunk 2]
```diff
@@ -333,6 +333,10 @@ def handle_operator(
     else:
         control_wire_inputs = control_values = ()
 
+    control_wire_inputs += self.control_wires
+    control_values += self.control_values
+    n_ctrls += len(self.control_wires)
```
Comment (paul0403) [inline frontend/catalyst/from_plxpr/qfunc_interpreter.py:338]: `self` is a jaxpr interpreter. Why would it have control wires?

File: frontend/test/lit/test_qref/test_flat_circuits.py [hunk 3]
```diff
@@ -70,6 +70,23 @@ def test_custom_op(i: int):
 print(test_custom_op.mlir)
 
 
+# CHECK-LABEL: func.func public @test_qfunc_ctrl_operator2_preserves_control_context
+@qp.qjit(capture=True, target="mlir")
+@qp.qnode(qp.device("null.qubit", wires=2))
+def test_qfunc_ctrl_operator2_preserves_control_context():
```
Comment (paul0403) [inline frontend/test/lit/test_qref/test_flat_circuits.py]: How is this test different than the other controlled gate tests in this file?

File: frontend/test/pytest/from_plxpr/test_from_plxpr.py [hunk 4]
```diff
@@ -437,13 +438,15 @@ def c():
         catalyst_xpr = from_plxpr(plxpr)()
         qfunc_xpr = catalyst_xpr.eqns[0].params["call_jaxpr"]
 
-        assert qfunc_xpr.eqns[-5].primitive == qref_qinst_p
+        assert qfunc_xpr.eqns[-5].primitive == qref_operator_p
```
Comment (paul0403) [inline frontend/test/pytest/from_plxpr/test_from_plxpr.py:441]: This is because migration of `S` to operator2 has been completed?

File: frontend/test/pytest/from_plxpr/test_from_plxpr.py [hunk 5]
```diff
@@ -484,6 +487,33 @@ def c(x, wire3):
...
+
+        plxpr = jax.make_jaxpr(circuit)()
+        catalyst_xpr = from_plxpr(plxpr)()
+        qfunc_xpr = catalyst_xpr.eqns[0].params["call_jaxpr"]
+
+        operator_eqn = next(
+            eqn
+            for eqn in qfunc_xpr.eqns
+            if eqn.primitive == qref_operator_p and eqn.params["op_cls"] is qp.S
+        )
```
Comment (paul0403) [inline frontend/test/pytest/from_plxpr/test_from_plxpr.py]: Like the other tests in this file, let's just grab the gate primitive from a direct indexing.

I don't like it either, but in my mind all the tests in this file should be lit tests instead of pytests and should be rewritten. For now, let's just be consistent.

File: frontend/catalyst/from_plxpr/qfunc_interpreter.py (hunk 2 above)
Comment (JerryChen97) [reply frontend/catalyst/from_plxpr/qfunc_interpreter.py:338]: `self` is the interpreter, but the interpreter instance also carries context while evaluating
nested JAXPRs. `handle_ctrl_transform()` creates `unroller = copy(self)`, appends the transform's
control wires and values to that copy, and evaluates the nested JAXPR with `unroller.eval(...)`.

Therefore, when an `operator_p` inside that nested JAXPR reaches `handle_operator()`, `self` is the
copied interpreter and `self.control_wires` and `self.control_values` contain the enclosing
quantum-function controls. The inner `operator_p` itself has `n_ctrls=0`, so the interpreter context
is the only carrier for those controls.

Constructed controlled operators encode their controls directly in the `operator_p` inputs. The
three lines in `handle_operator()` merge those controls with any enclosing quantum-function
controls before binding `qref_operator_p`. The legacy `interpret_operation()` path already performs
the same merge.

File: frontend/test/lit/test_qref/test_flat_circuits.py (hunk 3 above)
Comment (JerryChen97) [reply frontend/test/lit/test_qref/test_flat_circuits.py]: S is op2 now?

File: frontend/test/pytest/from_plxpr/test_from_plxpr.py (hunk 4 above)
Comment (JerryChen97) [reply frontend/test/pytest/from_plxpr/test_from_plxpr.py:441]: Yes

File: frontend/test/pytest/from_plxpr/test_from_plxpr.py (hunk 5 above)
Comment (JerryChen97) [reply frontend/test/pytest/from_plxpr/test_from_plxpr.py]: I agree. We should have another story to audit tests for post-Op2 era

File: frontend/test/lit/test_qref/test_flat_circuits.py (hunk 3 above)
Comment (paul0403) [reply frontend/test/lit/test_qref/test_flat_circuits.py]: But we already have dedicated operator 2 lowering lit tests when we added the lowering.

Comment (JerryChen97) [reply frontend/test/lit/test_qref/test_flat_circuits.py]: Ah... true, `test_doubly_ctrl(as_qfunc=True)` already covered this part. This lit test is indeed redundant

Comment (paul0403) [reply frontend/test/lit/test_qref/test_flat_circuits.py]: No I meant the lit tests in `test_operator.py` with the mocked out `Operator2` test classes

File: frontend/catalyst/from_plxpr/qfunc_interpreter.py (hunk 2 above)
Comment (paul0403) [reply frontend/catalyst/from_plxpr/qfunc_interpreter.py:338]: Ah, forgot contorl is a transform in plxpr

File: frontend/test/lit/test_qref/test_flat_circuits.py (hunk 3 above)
Comment (paul0403) [reply frontend/test/lit/test_qref/test_flat_circuits.py]: But anyway, feel free to leave this test in or delete it, up to you guys

File: frontend/test/pytest/from_plxpr/test_from_plxpr.py (hunk 5 above)
Comment (Qottmann) [reply frontend/test/pytest/from_plxpr/test_from_plxpr.py]: ✅

File: frontend/test/lit/test_qref/test_flat_circuits.py (hunk 3 above)
Comment (Qottmann) [reply frontend/test/lit/test_qref/test_flat_circuits.py]: deleted it ✅

**Review by paul0403 — APPROVED** [review]
